### ISSUE TITLE
Training with multiple datasets using the estimator

### ISSUE THREAD

**Issue by amitlevy21** [tensorflow/tensorflow#25285]
**System information**
- TensorFlow version (you are using): 1.12
- Are you willing to contribute it (Yes/No): Not in the near future

**Describe the feature and the current behavior/state.**

I'll start with what i'm trying to achieve.
I want to train a model (a language model) with Multitask Learning and Transfer Learning.
I want to train with such techniques while using 3 different tasks.
Each task has it's own loss function, optimizer method training hooks, evaluation hooks and different data input.
I'm using the tf.data TFRecord dataset.
I've read a lot about the head API which seems to solve my problem just as described in this image:

![1_jcoaugzhqn3_7oyvu58t4q](https://user-images.githubusercontent.com/28672915/51889143-512b2680-23a1-11e9-8756-f03c3522adbd.png)

but my problem is the features - the input data, which changes according to what task i'm training on.
I want to be able to control the data input with a single call to estimator.train or estimator.eval

**Current Behavior**

Suppose the following:
| Tasks  | Task's Dataset |
|--------|----------------|
| Task A | Database A     |
| Task B | Database B     |
| Task C | Database C     |

and i want to train one task for x steps and then perform a "context switch" - switch to a different task and train it for `x` steps as well.
So I train Task A using `estimator.train(input_fn=Dataset_A, steps=x)` and then i train Task B using `estimator.train(input_fn=Dataset_B, steps=x)`

The problem starts when i want to train Task A again for x steps. Task A will be trained be first x records from it's input_fn (Dataset_A), this occurs since that estimator recreates the input_fn (and model_fn). I know the estimator does it this by design but this causes issues when performing Multitask learning while training with different datasets.

The tf.data api supports this kind of functionality. Using the [feedable](https://www.tensorflow.org/guide/datasets#creating_an_iterator) iterator, i can switch between datasets by invoking a different string handles of each dataset i got. **But how can i make the estimator to manage different string handles of multiple datasets?**
I would like to invoke estimator.train once and have a control over the flow of the data during the training (maybe be using session run hooks?)

**Will this change the current api? How?**

In order to support feedable iterators the estimator should be able to expected an input_fn that returns multiple tf.data.Dataset objects or iterators of those datasets.

**Who will benefit with this feature?**

Anyone that will want to train and experiment models using multiple datasets.

**Comment by amitlevy21** [comment]
@karmel I'd love to hear your opinion on this. Did i provide enough info to describe the problem?

This might partly be related to #16087 and #18016. But i would like to emphasize that here I'm describing a use case that is need for the training process as well and that i would like to control the "context switch" of datasets with estimator hooks(tf.train.SessionRunHook).

Thank you.

**Comment by karmel** [comment]
Thanks, @amitlevy21 . I'm not totally sure I understand the use-case above, so let me repeat back to you what I hear: you want to switch between datasets, performing the same train operation on batches from each, but interleaving the steps. Is that correct? If so, Estimator doesn't interleave particularly well, but you can instead represent one input/model_fn step as a pass through each of the datasets-- ie, the input fn loads A, B, and C, and the model fn runs A, B, and C in one go. Each "step" is really a three-fold pass over the batch. Alternatively, the tf.keras subclassing API may be a better fit here, as it is not forced to be hermetic with each step in the way Estimator is.

**Comment by amitlevy21** [comment]
@karmel Thank you. You understand exactly what i'm trying to achieve.
I'll try my a go with the suggestions you made.

**Comment by svboeing** [comment]
Hello @amitlevy21. I am facing similar problem and want to ask you a question. You wrote 

> So I train Task A using estimator.train(input_fn=Dataset_A, steps=x) and then i train Task B using estimator.train(input_fn=Dataset_B, steps=x)

but you also mentioned that your tasks have different loss functions. How do you switch between them? As far as i understand, in this code snippet you only change your input_fn, but not model_fn and hence, model loss stays the same. I am facing similar issue as i want to implement a multi-task learning model, where different inputs from different datasets are processed through shared BERT encoder, on top of which there are task-specific layers and different task-specific losses. BERT instance is created inside my model_fn, where it gets called, and i only want one instance to be created. How do i make model_fn use the appropriate loss for current task, if i can not change the computational graph dynamically(e.g. pass task_id to model_fn to make it choose appropriate task loss by this id)? @karmel Your answer will also be greatly appreciated. 
Thank you

**Comment by amitlevy21** [comment]
@svboeing I'm implementing an infrastructure for conducting experiments on language models using multitask and transfer learning, my client have asked me to not use Keras implementation. The use of Keras, Especially with the upcoming Tensorflow 2.0 should have decrease the complexity of my code, So i suggest to only considerate the following if you must not use Keras in your solution. 

I used the follow function to create the train, validation and test data of each task:
```python
def _create_input_fn(tf_record_path, hyperparams):
    def input_fn():
        """
        This method is invoke each time we call estimator.train
        or estimator.eval or estimator.predict, the estimator recreates
        the tf.data.Dataset that we return here. For each dataset we train,
        we want to remember how many records we trained it so we can iterate
        different datasets each epoch or batch. Therefore, we create a dataset
        step counter for each new dataset that we load, and skip records from
        reloaded dataset according to it's step.
        Returns:
            tf.data.Dataset object representing the dataset
        """
        global dataset_step_counter

        logger.debug("TF records path=%s", tf_record_path)
        logger.debug("dataset_steps=%s", dataset_step_counter[tf_record_path])

        logger.info("training with dataset=%s", tf_record_path)

        shuffle = hyperparams.train.get_or_default(key="shuffle", default=False)
        shuffle_buffer_size = hyperparams.train.get_or_default(key="shuffle_buffer_size", default=10000)
        dataset = load_dataset(abs_tf_record_path=tf_record_path,
                               batch_size=hyperparams.train.batch_size,
                               feature_sample_size=hyperparams.data.shape_size_features,
                               label_sample_size=hyperparams.data.shape_size_labels,
                               skip_first_n=dataset_step_counter[tf_record_path],
                               shuffle=shuffle,
                               shuffle_buffer_size=shuffle_buffer_size)
        return dataset

    return input_fn

train_dataset = _create_input_fn(tf_record_path=train_tf_record_path,
                                     hyperparams=hyperparams)
validation_dataset = _create_input_fn(tf_record_path=valid_tf_record_path,
                                          hyperparams=hyperparams)
test_dataset = _create_input_fn(tf_record_path=test_tf_record_path,
                                    hyperparams=hyperparams)
```

I have a similar function, that returns a model_fn function each time its called:
```python
 estimator_spec = _create_tf_estimator_spec(create_model=create_model,
                                               create_loss=create_loss,
                                               create_optimizer=create_optimizer,
                                               shared_hyperparams=shared_hyperparams,
                                               training_hooks=training_hooks,
                                               evaluation_hooks=evaluation_hooks)
```

Where create_model, create_loss and create_optimizer are **functions**.
My code was inspired by this [talk](https://www.youtube.com/watch?v=Fudxo6WZHO0&t=1s) which i recommend.

So basically, i am able to train different tasks where each has different a different dataset, loss and optimizer by invoking ```_create_input_fn``` and ```_create_tf_estimator_spec``` for each task.
the create_model method must be the same for all tasks, which means it contains a logits layer for each of the tasks, which are used in the create_loss function.

There are some issues with this approach, which are derived from the fact that the estimator recreates by design the input_fn and model_fn each time you call ```estimator.train```. Which is why i had to use ```dataset_step_counter``` global variable that recalls how many records the model has seen from dataset X and then skip those records the next time ```estimator.train``` is called.

Another main issue  is the use of train and evaluation hooks which will not preserve state between calls of ```estimator.train```.

Again, the use of Keras will simplify a lot of the code and reduce the issues you will have if you will only use the estimator.

**Comment by zjh-nudger** [comment]
I am also facing similar problem and want to ask question.

In your implementation, how to share specific module between different task?

thank you

**Comment by zjh-nudger** [comment]
> @svboeing I'm implementing an infrastructure for conducting experiments on language models using multitask and transfer learning, my client have asked me to not use Keras implementation. The use of Keras, Especially with the upcoming Tensorflow 2.0 should have decrease the complexity of my code, So i suggest to only considerate the following if you must not use Keras in your solution.
> 
> I used the following function to create the train, validation and test data of each task:
> 
> ```python
> def _create_input_fn(tf_record_path, hyperparams):
>     def input_fn():
>         """
>         This method is invoke each time we call estimator.train
>         or estimator.eval or estimator.predict, the estimator recreates
>         the tf.data.Dataset that we return here. For each dataset we train,
>         we want to remember how many records we trained it so we can iterate
>         different datasets each epoch or batch. Therefore, we create a dataset
>         step counter for each new dataset that we load, and skip records from
>         reloaded dataset according to it's step.
>         Returns:
>             tf.data.Dataset object representing the dataset
>         """
>         global dataset_step_counter
> 
>         logger.debug("TF records path=%s", tf_record_path)
>         logger.debug("dataset_steps=%s", dataset_step_counter[tf_record_path])
> 
>         logger.info("training with dataset=%s", tf_record_path)
> 
>         shuffle = hyperparams.train.get_or_default(key="shuffle", default=False)
>         shuffle_buffer_size = hyperparams.train.get_or_default(key="shuffle_buffer_size", default=10000)
>         dataset = load_dataset(abs_tf_record_path=tf_record_path,
>                                batch_size=hyperparams.train.batch_size,
>                                feature_sample_size=hyperparams.data.shape_size_features,
>                                label_sample_size=hyperparams.data.shape_size_labels,
>                                skip_first_n=dataset_step_counter[tf_record_path],
>                                shuffle=shuffle,
>                                shuffle_buffer_size=shuffle_buffer_size)
>         return dataset
> 
>     return input_fn
> 
> train_dataset = _create_input_fn(tf_record_path=train_tf_record_path,
>                                      hyperparams=hyperparams)
> validation_dataset = _create_input_fn(tf_record_path=valid_tf_record_path,
>                                           hyperparams=hyperparams)
> test_dataset = _create_input_fn(tf_record_path=test_tf_record_path,
>                                     hyperparams=hyperparams)
> ```
> 
> I have a similar function, that returns a model_fn function each time its called:
> 
> ```python
>  estimator_spec = _create_tf_estimator_spec(create_model=create_model,
>                                                create_loss=create_loss,
>                                                create_optimizer=create_optimizer,
>                                                shared_hyperparams=shared_hyperparams,
>                                                training_hooks=training_hooks,
>                                                evaluation_hooks=evaluation_hooks)
> ```
> 
> Where create_model, create_loss and create_optimizer are **functions**.
> My code was inspired by this [talk](https://www.youtube.com/watch?v=Fudxo6WZHO0&t=1s) which i recommend.
> 
> So basically, i am able to train different tasks where each has different a different dataset, loss and optimizer by invoking `_create_input_fn` and `_create_tf_estimator_spec` for each task.
> the create_model method must be the same for all tasks, which means it contains a logits layer for each of the tasks, which are used in the create_loss function.
> 
> There are some issues with this approach, which are derived from the fact that the estimator recreates by design the input_fn and model_fn each time you call `estimator.train`. Which is why i had to use `dataset_step_counter` global variable that recalls how many records the model has seen from dataset X and then skip those records the next time `estimator.train` is called.
> 
> Another main issue is the use of train and evaluation hooks which will not preserve state between calls of `estimator.train`.
> 
> Again, the use of Keras will simplify a lot of the code and reduce the issues you will have if you will only use the estimator.

I am also facing similar problem and want to ask question.

In your implementation, how to share specific module between different task?

thank you

**Comment by amitlevy21** [comment]
@zjh-nudger Did you mean a python module?
Each task is a module that has a file for each of the following functions: create_loss, create_model, create_optimizer.
Then i can simply import that of the those functions anywhere.

**Comment by agupta74** [comment]
> There are some issues with this approach, which are derived from the fact that the estimator recreates by design the input_fn and model_fn each time you call estimator.train. Which is why i had to use dataset_step_counter global variable that recalls how many records the model has seen from dataset X and then skip those records the next time estimator.train is called.

@amitlevy21 It looks like you solved the issue of recreation of input_fn by using a global data_step_counter variable. How did you solve the issue of recreating graph when model_fn gets called every time estimator.train is called?

**Comment by amitlevy21** [comment]
@agupta74 which issue exactly? I'll guess you're talking about preserving the trained weights?
```python
    config = tf.estimator.RunConfig(model_dir=checkpoint_path,
                                    save_summary_steps=summary_steps,
                                    save_checkpoints_steps=save_checkpoint_steps,
                                    keep_checkpoint_max=keep_checkpoints_max)

```
Define a run config pointing the same model_dir. The same model_dir gets called each time i call train on the estimator so it does not matter to me that the graph is recreated again and again.

**Comment by agupta74** [comment]
@amitlevy21 Thanks for your quick response !  Yes, I meant how do you preserve the trained weights for multiple estimator.train calls. So it looks like we would have to save checkpoints on disk at the end of every estimator.train call before we can call estimator.train on the next task. I guess this approach would be very slow for use cases with multi-task learning where the task type changes after every k steps where k = 1.

**Comment by amitlevy21** [comment]
@agupta74 actually I have performance issues when k > 1. Since calling estimator.train also looks for available GPU devices. Not sure why, but the training speed decreases over time when i conduct more than 4 experiments, each of them performing multitask training on 2-3 tasks. 
Again, i think there might be a better design for multitask learning purposes using Keras.

**Comment by strubell** [comment]
It seems like [tf.data.experimental.choose_from_datasets](https://www.tensorflow.org/api_docs/python/tf/data/experimental/choose_from_datasets) might now provide this functionality?

**Comment by amitlevy21** [comment]
@strubell I have tried using it for my use case, but i needed a much more flexible way to interleave between the datasets. I also took a look at [tf.data.Dataset.interleave](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#interleave) but struggled a lot for two main reasons:
1. How can i identify which task am i going to train and load its dataset accordingly?
2. How can i interleave between datasets themselves and tasks themselves and apply 1? 

Using tf.data.experimental.choose_from_dataset and tf.data.Dataset.interleave may solved many use cases. But i could not make it to fit my use case.

**Comment by strubell** [comment]
choose_from_datasets takes two parameters: a list of datasets (in your case this would be a list of datasets corresponding to all the tasks you want to train on) and a list of indices into that list (this determines, for a given training step, which task/dataset the batch comes from). The list of indices achieves interleaving. If the problem is knowing which data you're dealing with in the model, I imagine you could pass some indicator in with the data that tells you which task it's for, then use `tf.cond` to execute different tf code depending on the value of that variable.
